src/concurrency.h

```diff
@@ -8,140 +8,178 @@
 #include <thread>
 #include <atomic>
 #include <functional>
+#include <cassert>
 
 const int QUEUE_MAX_SIZE = 10;
 // Simplest implementation of a blocking concurrent queue for thread messaging
 template<class T>
 class single_consumer_queue
 {
     std::deque<T> _queue;
-    std::mutex _mutex;
+    mutable std::mutex _mutex;
```

Comment:
why it should be mutable?

Reply:
So we can use `const` function declarations

Reply:
but there is no const function...

Reply:
size(), empty()

Reply:
but they don't use the mutex

Reply:
they don't use the mutex

Reply:
```
    size_t size() const
    {
        std::lock_guard< std::mutex > lock( _mutex );
        return _queue.size();
    }
```

What's the disconnect?

Reply:
ok didn't see it